addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java

```diff
@@ -0,0 +1,403 @@
+        Map<String, UniFiClient> clientMap = new HashMap<String, UniFiClient>();
+        List<UniFiClient> clients = response.getData();
+        logger.debug("Found {} UniFi Client(s):", clients.size());
+        for (UniFiClient client : clients) {
+            client.setDevice(devicesCache.get(client.getDeviceMac()));
+            logger.debug("  {}", client);
+            clientMap.put(client.getMac(), client);
+        }
+        return clientMap;
+    }
+
+    private Map<String, UniFiClient> getClients() throws UniFiException {
+        Map<String, UniFiClient> clientMap = Collections.emptyMap();
+        Collection<UniFiSite> sites = sitesCache.values();
+        if (!sites.isEmpty()) {
+            clientMap = new HashMap<String, UniFiClient>();
+            for (UniFiSite site : sites) {
+                Map<String, UniFiClient> siteClientMap = getClients(site);
+                if (siteClientMap != null) {
+                    clientMap.putAll(siteClientMap);
+                }
+            }
+        }
+        return clientMap;
+    }
+
+    private Map<String, UniFiClient> getInsights(UniFiSite site) throws UniFiException {
+        UniFiClientResponse response = get(getInsightsUrl(site), INSIGHTS_WITHIN_24H, UniFiClientResponse.class);
+        Map<String, UniFiClient> insightsMap = new HashMap<String, UniFiClient>();
+        List<UniFiClient> clients = response.getData();
+        logger.debug("Found {} UniFi Insights(s):", clients.size());
+        for (UniFiClient client : clients) {
+            logger.debug("  {}", client);
+            insightsMap.put(client.getMac(), client);
+        }
+        return insightsMap;
+    }
+
+    private Map<String, UniFiClient> getInsights() throws UniFiException {
+        Map<String, UniFiClient> insightsMap = Collections.emptyMap();
+        Collection<UniFiSite> sites = sitesCache.values();
+        if (!sites.isEmpty()) {
+            insightsMap = new HashMap<String, UniFiClient>();
+            for (UniFiSite site : sites) {
+                Map<String, UniFiClient> siteInsightsMap = getInsights(site);
+                if (siteInsightsMap != null) {
+                    insightsMap.putAll(siteInsightsMap);
+                }
+            }
+        }
+        return insightsMap;
+    }
+
+    private boolean belongsToSite(UniFiClient client, String siteName) {
+        boolean result = true; // mgb: assume true = proof by contradiction
+        if (StringUtils.isNotEmpty(siteName)) {
+            UniFiSite s = sitesCache.get(client.getSiteId());
+            // mgb: if the 'site' can't be found or the name doesn't match...
+            if (s == null || !s.matchesName(siteName)) {
+                // mgb: ... then the client doesn't belong to this thing's configured 'site' and we 'filter' it
+                result = false;
+            }
+        }
+        return result;
+    }
+
+    // Public API functions
```

Comment:
I would expect those below the private methods (on which they depend)

Reply:
Ahum, I hope you read this as above :-S

Reply:
So you mean above?

Just to be clear...
```
// Public API

public foo() {
     bar();
}

// Private API

private bar() {
    logger.debug("martinv needs to make up his mind");
}
```

🙊

Reply:
😁

Yes first public methods, but they don't have to be in blocks. Having a public method immediately followed by its private helper methods is also fine.

But if you use blocks, public should on top after your constructors.